OAuth2 support

kuzmany opened the pull request:
This PR brings OAuth2 support

Documentation: https://github.com/mautic/documentation/pull/375

kuzmany committed:
Oauth2 for webmecanik

kuzmany committed:
Add base url to connectionLabel

kuzmany committed:
Add help text to clientId and clientSecret

kuzmany committed:
Unit test reworked

npracht commented:
**To give a bit more of context: we've been working with @kuzmany to enhance the Zapier plugin.**

Today, it is using basic auth which is.... not nice it terms of security.
**With this PR we suggest the move to OAuth2 authentification offered by mautic app.**

escopecz commented:
Can this be done without the BC break? If we release this then the Zapier users will have to go and reauthenticate all their Zaps. Ideally, it would be great if the integration would still support basic auth for existing Zaps and Oauth2 for new ones.

npracht commented:
Or if not possible, create a V2 plugin, i've seen this behavior often.

kuzmany commented:
@escopecz I think It works like you describe. 
The current Zaps will use old version and new Zaps will use newer.
Then we can continue maintenance both version, or just new one (I prefer just new one - Oauth2 is more secure).

@npracht Can you send me example of V2 plugin. Because what I've read from docs, they allow just one public version for one brand. 
We could try releaase Mautic Oauth2, but I am not sure if they approve it.

npracht commented:
Then that means they changed their policy (which can be understandable).
@kuzmany can you ask their support concerning the fact that current Zaps are not broken when we deliver this kind of update ?

kuzmany commented:
@npracht Based on docs, not broken for them. If you use Zap for Mautic Zapier 2.16 then you stay on that version forever. You can migrate that user in dev platform, but not necessary.

escopecz commented:
We can have Zapier users on 2 versions simultaneously for short period of time. We can even announce (I think, never done that) that hey will have to reauthenticate if they want to migrate to the newer version. And schedule the old version to disappear after 6 months or so. But there is still the need for hundreds of users to do some manual work with low value (from basic user standpoint) they gain for it. I would like to avoid that if possible. This may be another topic to poll the community about.

kuzmany committed:
Revert back results.should.not.be.empty()

npracht commented:
Let do it like that. @escopecz there is a valuable reason: security. Basic Auth should not exists.

escopecz commented:
@kuzmany we'll also need an update to https://github.com/mautic/mautic-zapier/blob/master/CHANGELOG.md for the release. And the version bump in https://github.com/mautic/mautic-zapier/blob/master/package.json#L3

escopecz commented on authentication.js:
```diff
@@ -1,30 +1,120 @@
+    // This could also use the request shorthand.
+    getAccessToken: getAccessToken,
+    // (Optional) If the access token expires after a pre-defined amount of time, you can implement
+    // this method to tell Zapier how to refresh it.
```
These 2 lines of comment are not suited for production. Maybe go through all the comment and make it relevant. I think they were copy-pasted from some tutorial.

escopecz commented on test/authentication.js:
```diff
@@ -1,131 +1,33 @@
-      });
-  }).timeout(15000);
 
-  it('fails auth on not existent url', (done) => {
```
Can we keep this test?

escopecz replied:
@kuzmany and this one?

escopecz commented on test/authentication.js:
```diff
@@ -1,131 +1,33 @@
-      });
-  }).timeout(15000);
-
-  it('fails auth on invalid url', (done) => {
```
Can we keep this test?

escopecz replied:
The reason is that during the very long and detail review of this integration before Zapier made it public they insisted on useful error messages for users who do something wrong. So we should test for these cases.

escopecz replied:
@kuzmany did you see this comment?

kuzmany replied:
I don't know what do you expect from that tests. If oauth2 going to invalid url, then request failed always.
I've added fails auth on invalid url test, but let me know what do you want to add exactly.

escopecz reviewed (changes requested):
Can you also update Readme with how to test now? For example how to generate the access token?

kuzmany committed:
Add refresh_token to refreshAccessToken method

kuzmany committed:
Rewrite refreshAccessToken

kuzmany committed:
Improve comments

kuzmany committed:
Change test hook

kuzmany committed:
add refresh_token to store procedure

kuzmany committed:
Update README.md

kuzmany committed:
auth fix again

kuzmany committed:
Bump version to 3.0.0

kuzmany committed:
test: fails auth on invalid url

kuzmany committed:
Fix issue based on validation of project

polling sample in Zap includes fields not found in latest task history ; extra keys: "dateModified", "lastActive", "modifiedBy", "modifiedByUser" (T006)

kuzmany committed:
Update .gitignore

for PHPStorm

RCheesley commented:
I think that it would be best if we were to have a v2 of the app, and make it clear to folks when we are going to deprecate the v1.

That way it does not immediately break things but allows people the time to transition over to the new version.  

Does that make sense?

escopecz committed:
Adding a script that will generate the access token to run the tests suite

escopecz commented:
@kuzmany I tested this PR, but the majority of the automated functional tests are failing. I've created a script to get the access token (see https://github.com/mautic/mautic-zapier/pull/18/commits/6fc918dc7dd023f0f0521ff589bc7e7bb33452e2) to make the testing faster. Are the tests passing for you?

kuzmany commented:
@escopecz when I build it, test passed (without it we were not able to upload it ). 
I can check it again.
